code/modules/GOAP/GOAPInfoprovider.dm

```diff
@@ -0,0 +1,94 @@
+
+//This type is the eyes/ears of a GOAPAgent
+//There aren't many suitable defaults, so you must override these procs
+//To get what you want
+
+/datum/goap_info_provider
+	var/list/failures_past //assoc list of "failed goal state" = "world state at the time it failed"
+	//both of these are stored as "goal HASH" = failure_state object, since we're matching contents of lists, not the lists themselves
+	//Used to ensure we don't try again if the world hasn't changed, mainly for performance
+
+/datum/goap_info_provider/New()
+	failures_past = list()
+
+
+//Returns an associative of "statekey" = someval
+//Someval should be a boolean or comparable object
+//BEcause all comparisons between states are == (so no >= or > or < yet, sadly)
+/datum/goap_info_provider/proc/GetWorldState(datum/goap_agent/agent)
+	. = list()
+	if(agent.agent)
+		if(ismob(agent.agent))
```

Comment:
you don't need the agent.agent check, null is obviously not a mob so it would fail the second check anyway